### PR TITLE
[5.1] Update str_slug to convert "&" to "and", and also allow other custom replacements

### PR DESCRIPTION
**Pull request by ryanwinchester**
It would be nice if when I _sluggified_ `Friends & Family` it came out as `friends-and-family` instead of `friends-family`

Behaviour after commit:

``` php
$string = "Friends & Family";

$slug = str_slug($string); // 'friends-and-family'
```

And, just for fun:

``` php
$string = "Hello || Goodbye";

$replacements = [
    '||' => 'or',
]

$slug = str_slug($string, '-', $replacements); // 'hello-or-goodbye'
```

could also be used to adjust for other languages, or whatever:

``` php
$string = "Bonjour & Au revoir";

$replacements = [
    '&' => 'et',
]

$slug = str_slug($string, '-', $replacements); // 'bounjour-et-au-revoir'
```

### CODE REVIEW

File: src/Illuminate/Support/Str.php [hunk 1]
```diff
@@ -279,14 +279,19 @@ public static function singular($value)
...
+	 * @param  string $separator
+	 * @param array $replacements
 	 * @return string
 	 */
-	public static function slug($title, $separator = '-')
+	public static function slug($title, $separator = '-', $replacements = array())
 	{
 		$title = static::ascii($title);
 
+		$replacements['&'] = isset($replacements['&']) ? $replacements['&'] : 'and';
```
Comment (Arrilot) [inline src/Illuminate/Support/Str.php]: I would probably prefer moving this to function argument to make it disableable
`public static function slug($title, $separator = '-', $replacements = ['&' => 'and'])`

Comment (fungku) [inline src/Illuminate/Support/Str.php]: I like that way too, but then you need to re-add that one whenever you add other custom replacements....  I was torn between which way is better...

File: src/Illuminate/Support/Str.php [hunk 2]
```diff
@@ -279,14 +279,19 @@ public static function singular($value)
 	/**
 	 * Generate a URL friendly "slug" from a given string.
 	 *
-	 * @param  string  $title
-	 * @param  string  $separator
+	 * @param  string $title
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/Str.php]: cs the cs

Comment (GrahamCampbell) [inline src/Illuminate/Support/Str.php]: fix the cs please

File: src/Illuminate/Support/Str.php [hunk 3]
```diff
@@ -279,14 +279,19 @@ public static function singular($value)
...
+	 * @param  string $separator
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/Str.php]: Clarkson

File: src/Illuminate/Support/Str.php [hunk 4]
```diff
@@ -279,14 +279,19 @@ public static function singular($value)
...
+	 * @param array $replacements
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/Str.php]: Clarkson

File: src/Illuminate/Support/Str.php (hunk 3 above)
Comment (GrahamCampbell) [inline src/Illuminate/Support/Str.php]: fix the cs please

File: src/Illuminate/Support/Str.php (hunk 4 above)
Comment (GrahamCampbell) [inline src/Illuminate/Support/Str.php]: fix the cs please

File: src/Illuminate/Support/helpers.php [hunk 5]
```diff
@@ -709,13 +709,14 @@ function str_singular($value)
 	/**
 	 * Generate a URL friendly "slug" from a given string.
 	 *
-	 * @param  string  $title
-	 * @param  string  $separator
+	 * @param  string $title
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/helpers.php]: fix the cs please

File: src/Illuminate/Support/helpers.php [hunk 6]
```diff
@@ -709,13 +709,14 @@ function str_singular($value)
...
+	 * @param  string $separator
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/helpers.php]: fix the cs please

File: src/Illuminate/Support/helpers.php [hunk 7]
```diff
@@ -709,13 +709,14 @@ function str_singular($value)
...
+	 * @param  array $replacements
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/helpers.php]: fix the cs please

**Comment by TheShiftExchange** [comment]
This should be noted as a potential breaking change.

People might be relying on the current expected behavior to generate URLs, filenames etc.

Suddenly having different output from a helper could easily lead to issues..

(Note - I'm not against the change in principle - but it needs to be against at least 5.1)

**Comment by Arrilot** [comment]
It is against 5.1 already.

**Comment by TheShiftExchange** [comment]
@Arrilot - yeah, sorry - misread the fork name.

It will need to be mentioned in the Upgrade 5.1 docs though - so people are aware of the change in behavior.

**Comment by taylorotwell** [comment]
I don't really want to make this breaking change as some people could be using those slugs as look-ups and it doesn't add a lot of new behavior.

**Comment by fungku** [comment]
@taylorotwell what if replacement doesn't default to anything, so the change is unbreaking, and then at least there's the option to add replacements?

**Pull request closed without merging**